src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1704,19 +1704,13 @@ private int tryExecute(final Runnable runnable) {
         if (TAIL_LOCK) lockTail();
         TaskNode tail = this.tail;
         TaskNode node = null;
+        int spins = 0;
         for (;;) {
             tailNext = tail.getNext();
             if (tailNext instanceof TaskNode) {
-                TaskNode tailNextTaskNode;
-                do {
-                    if (UPDATE_STATISTICS) spinMisses.increment();
-                    tailNextTaskNode = (TaskNode) tailNext;
-                    // retry
-                    tail = tailNextTaskNode;
-                    tailNext = tail.getNext();
-                } while (tailNext instanceof TaskNode);
-                // opportunistically update for the possible benefit of other threads
-                if (UPDATE_TAIL) compareAndSetTail(tail, tailNextTaskNode);
+                onSpinMisses(spins);
```

Comment:
I think we need to update spins here
```suggestion
                spins = onSpinMisses(spins);
```

I kicked off a benchmark run with a very similar change to this one a half hour ago, though it still has about three more hours to go. https://github.com/carterkozak/jboss-threads/commit/5bfc23377ad0406d5b5443a8847fb2b739959182

I imagine that including onSpinWait/yield in other paths will be helpful as well, I can try to test that once this round completes. It may be worth testing with a similar structure for consumers in getOrAddNode.

Reply:
Results from that run on https://github.com/carterkozak/jboss-threads/commit/5bfc23377ad0406d5b5443a8847fb2b739959182 are available here:
https://gist.githubusercontent.com/carterkozak/3166809eeed4267c1b590f2afe92be80/raw/4caacdf55a45cacec499a88a67356ef0ecac4a36/jboss-perf.log

Some results using spinWaitCompletion appear to be dominated by the spinning waiter.

the 2-core 16,16 benchmarks were previously the worst case for no-locks by a wide margin, but using this change the results appear to be  equivalent!

```
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16  LTQ_THREAD_POOL_EXECUTOR                 false  avgt    4     70.660 ±    50.534  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16  LTQ_THREAD_POOL_EXECUTOR                  true  avgt    4    110.110 ±    29.931  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16  LBQ_THREAD_POOL_EXECUTOR                 false  avgt    4     92.357 ±     6.242  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16  LBQ_THREAD_POOL_EXECUTOR                  true  avgt    4  13535.035 ±  7852.442  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16              EQE_NO_LOCKS                 false  avgt    4     60.315 ±     8.330  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16              EQE_NO_LOCKS                  true  avgt    4    784.051 ±   272.482  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16             EQE_SPIN_LOCK                 false  avgt    4     64.053 ±    48.213  us/op
ExecutorBenchmarks.benchmarkSubmit            1        2              16,16             EQE_SPIN_LOCK                  true  avgt    4    781.773 ±   395.615  us/op
```

Tests using 100,200 will likely benefit from the additional onSpinWait/yields. I'll kick off another run from this PR.

Reply:
> Some results using spinWaitCompletion appear to be dominated by the spinning waiter.

Yep, indeed it's a double edged sword: but sometime makes more evident the real latency, depending by the timing where Future::get is hit. I still think it worths to run benchs checking both to acknowledge when it happens

Thanks about checking with this PR, I think as well that th additional onSpinWait/yields are helpful under heavy load

Reply:
It's strange anyway, I gotten exactly the opposite results on my Xeon :O (spinWaitCompletion being much faster with 2 cores).

Reply:
That’s odd! I updated the benchmark to use onSpinWait instead of yield, that could be the problem. Perhaps it should yield every n spins for this case?

Reply:
Results from this PR on 5165fc11077223580fc29d30762fb36da6993886

https://gist.githubusercontent.com/carterkozak/3d7e43337b92f180364d2b5800b18f3a/raw/f97e4ba4584323dc8814afa8e3c2a110342fd3d0/jboss-perf-5165fc11077223580fc29d30762fb36da6993886.log

I haven't had a chance to do a comparison yet, just posting raw data :-)

Reply:
The results look great to me and I think that's time to remove the LOCK_TAIL code path, given that it seems no longer necessary: re your box; I see a great variance, while I'm not getting that much variant on mine...are you sure is a busy one? any form of turbo boost is disabled?

Reply:
While I haven't disabled any sort of frequency scaling in the bios, I'm using the performance cpu governor which should keep frequencies relatively steady. Over the three hour benchmark I wouldn't be surprised if there was some thermal throttling, unfortunately I'm not recording historical temperatures at the moment.